api.c

```diff
@@ -0,0 +1,214 @@
+/* Pi-hole: A black hole for Internet advertisements
+*  (c) 2017 Pi-hole, LLC (https://pi-hole.net)
+*  Network-wide ad blocking via your own hardware.
+*
+*  FTL Engine
+*  General API commands
+*
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license. */
+
+#include "FTL.h"
+#include "api.h"
+#include "cJSON.h"
+
+void sendAPIResponse(int sock, char type, char http_code) {
+	sendAPIResponseWithCookie(sock, type, http_code, NULL);
+}
+
+void sendAPIResponseWithCookie(int sock, char type, char http_code, const long *session) {
+	char *http_status;
+
+	switch(http_code) {
+		default:
+		case OK:
+			http_status = "200 OK";
+			break;
+		case BAD_REQUEST:
+			http_status = "400 Bad Request";
+			break;
+		case INTERNAL_ERROR:
+			http_status = "500 Internal Server Error";
+			break;
+		case NOT_FOUND:
+			http_status = "404 Not Found";
+			break;
+		case UNAUTHORIZED:
+			http_status = "401 Unauthorized";
+			break;
+	}
+
+	// Send header only for full HTTP requests
+	if(type == APIH)
+	{
+		if(session == NULL) {
+			// No cookie to send
+			ssend(sock,
+			      "HTTP/1.0 %s\nServer: FTL\nCache-Control: no-cache\nAccess-Control-Allow-Origin: *\n"
+					      "Content-Type: application/json\n\n{", http_status);
+		}
+		else {
+			// Send cookie
+			ssend(sock,
+			      "HTTP/1.0 %s\nServer: FTL\nCache-Control: no-cache\nAccess-Control-Allow-Origin: *\n"
+					      "Set-Cookie: FTL_SESSION=%ld\nContent-Type: application/json\n\n{", http_status, *session);
+		}
+	}
+}
+
+// session will have the client's valid session written to, if it's not unauthorized
+enum Auth authenticate(char *with_headers, char *payload, long *session, int sock) {
+	// First figure out if the client has authenticated before.
+	char *sessionStr;
+	AuthData *auth = NULL;
+
+	// Find the cookie header (will contain a long int value)
+	if(strstr(with_headers, "Cookie: ") != NULL && (sessionStr = strstr(with_headers, "FTL_SESSION=")) != NULL) {
```

Comment:
This looks dangerous. Doesn't this read `FTL_SESSION=` from anywhere in the http request? I can't think of how this would be an attack vector on its own, but coupled with other things, it probably could be. Maybe.